NETOBSERV-2045: Create packets xlation blog for netobserv (#3)
netobserv/netobserv.github.io: content/posts/2025-01-03-packets-xlation-enrichment.md

Kubernetes service flows, providing insights into backend behavior and improving observability.

## The Challenge

When traffic flows through a Kubernetes Service, it often gets abstracted away by kube-proxy or other load balancers.
For example, a request to my-service on ClusterIP is transparently routed to one of the available pods.
However, traditional flow logs might only capture the virtual service endpoint rather than the actual pod IP
that served the request. This lack of granularity makes troubleshooting and optimization more challenging.

To solve this, we can:

Capture the network flows at the kernel level.

Enrich them with translated endpoint information, showing not only the service but also the specific backend pod.

## How eBPF Can Help

eBPF allows you to execute custom programs in the Linux kernel, making it an ideal tool for network observability.
By attaching eBPF programs to network events, you can capture and enrich data about service flows.

Here’s a high-level approach:

Attach eBPF Programs:
Use eBPF programs to hook into kernel networking events, such as kprobe on functions like `nf_nat_manip_pkt`.

Enrich Flow Logs:

> As network packets are processed, augment flow logs with metadata about the translated endpoint. This can include:

Author: that is what ebpf code will do user doesn't need to do anything

- Source Pod IP/Name

- Source Port

- Destination Pod IP/Name

- Destination Port

- Conntrack Zone ID

- ICMP code

Benefits of Using eBPF

Granular Observability: Directly see which pod served a request.

Low Overhead: Operates in the kernel with minimal impact on performance.

Dynamic Updates: Respond to changes in Kubernetes without requiring application restarts.

Simplified Architecture: No need for sidecars or intrusive network plugins.

## How to enable Packet Translation enrichment feature

 To enable packet translation enrichment feature, need to create a flow collector with the following feature enabled

```yaml
apiVersion: flows.netobserv.io/v1beta2
kind: FlowCollector
metadata:
  name: cluster
spec:
  agent:
    type: eBPF
    ebpf:
      features:
      - "PacketTranslation"
```

## Example

>  To demonstrate this feature, let's configure a `CluesterIP` kubernetes service using the following example yaml

Reviewer: extra e in `ClusterIP`

```yaml
---
apiVersion: v1
kind: Pod
metadata:
 name: hello-pod
 namespace: xlat-test
 labels:
  app: hello-pod
spec:
 containers:
 - name: hello-world
   image: gcr.io/google-samples/node-hello:1.0
   ports:
   - containerPort: 8080
     protocol: TCP
---
kind: Service
apiVersion: v1
metadata:
 name: hello-pod
 namespace: xlat-test
spec:
  ports:
  - name: http
    port: 80
    protocol: TCP
    targetPort: 8080
  selector:
    app: hello-pod
  type: ClusterIP
```

> check the created service to find the `CluterIP` and `Port`

Reviewer: missing `s` in `CluterIP`

```bash
$oc describe svc -n xlat-test
Name:              hello-pod
Namespace:         xlat-test
Labels:            <none>
Annotations:       <none>
Selector:          app=hello-pod
Type:              ClusterIP
IP Family Policy:  SingleStack
IP Families:       IPv4
IP:                172.30.165.151
IPs:               172.30.165.151
Port:              http  80/TCP
TargetPort:        8080/TCP
Endpoints:         10.129.0.37:8080
Session Affinity:  None
Events:            <none>

$oc get pods -n xlat-test
oc get pods -n xlat-test
NAME        READY   STATUS    RESTARTS   AGE
client      1/1     Running   0          5s
hello-pod   1/1     Running   0          8m54s

```

> Then we can send traffic to this service IP and check the enriched flows on the netobserv console

Reviewer 2: - suggest you vs we to make it more customer-centric.
- I think since we refer to the console here, best to call it network observability since that whole name will be spelled out in the console. 
- add a colon to introduce the code.

```bash
$ while true; do oc exec -i -n xlat-test client -- curl 172.30.165.151:80 ; sleep 1; done
```

> from Netobserv console let's filter on Traffic destination `Kind` is `Service` in `xlat-test` namespace

Reviewer 2: ```suggestion
4. From the network observability console **Network Traffic** page, click the *Taffic flows* tab  and filter on Traffic destination `Kind` is `Service` in the `xlat-test` namespace:
```

Reviewer 2: I suggest clarifying the exact area of the console where they'll need to be working.

<img src="{page.image('packets-xlation-enrichment/pkt-xlat.png')}" alt="Service Traffic's Packet Translation enrichment using POD's names">

<img src="{page.image('packets-xlation-enrichment/pkt-xlat-ip-port.png')}" alt="Service Traffic's Packet Translation enrichment using POD's IP and Port">

> The following shows possible packet translation columns options, currently `zoneid` is the only visible column by default

Reviewer 2: ```suggestion
The following shows possible packet translation columns options. Currently `zoneid` is the only visible column by default:
```

Let's configure a `ClusterIP` Kubernetes service to try this feature!

1- Configure a `ClusterIP` Kubernetes service using the following example yaml:

```yaml
---
apiVersion: v1
kind: Pod
metadata:
 name: hello-pod
 namespace: xlat-test
 labels:
  app: hello-pod
spec:
 containers:
 - name: hello-world
   image: gcr.io/google-samples/node-hello:1.0
   ports:
   - containerPort: 8080
     protocol: TCP
---
kind: Service
apiVersion: v1
metadata:
 name: hello-pod
 namespace: xlat-test
spec:
  ports:
  - name: http
    port: 80
    protocol: TCP
    targetPort: 8080
  selector:
    app: hello-pod
  type: ClusterIP
```

2- Check the created service to find the `ClusterIP` and `Port`:

```bash
$oc describe svc -n xlat-test
Name:              hello-pod
Namespace:         xlat-test
Labels:            <none>
Annotations:       <none>
Selector:          app=hello-pod
Type:              ClusterIP
IP Family Policy:  SingleStack
IP Families:       IPv4
IP:                172.30.165.151
IPs:               172.30.165.151
Port:              http  80/TCP
TargetPort:        8080/TCP
Endpoints:         10.129.0.37:8080
Session Affinity:  None
Events:            <none>

$oc get pods -n xlat-test

> oc get pods -n xlat-test

Reviewer: Duplicate line not needed I think

tags: eBPF,Services,Monitoring,Troubleshooting
authors: [msherif1234]
---

# Enhancing NetObserv for Kubernetes Service Flows using eBPF: Adding Translated Endpoint Information

## Introduction

In a cloud-native environment, Kubernetes has become the de facto standard for managing containerized workloads.
While its service discovery and load-balancing features are robust, gaining visibility into the actual endpoints
serving traffic often requires complex instrumentation or external tooling.
This is where eBPF (extended Berkeley Packet Filter) shines, offering a way to deeply observe and enrich Kubernetes
service flows without intrusive changes to your applications.

In this blog, we'll explore how eBPF can be used to add translated endpoint (xlated endpoint) information to
Kubernetes service flows, providing insights into backend behavior and improving observability.

## The Challenge

When traffic flows through a Kubernetes Service, it often gets abstracted away by kube-proxy or other load balancers.
For example, a request to `my-service` on `ClusterIP` is transparently routed to one of the available pods.

> However, network observability will capture both the service traffic and the virtual service endpoint as two separate flows.

Reviewer 3: Behind the scene it's more than network observability issue here right ?
Any observability tool based on ebpf will report this. Maybe expose that instead

Author: not all observability tools are doing flow based capturing like us ? I guess what I meant to say here there is no association between svc traffic and its endpoint traffic user have to do this correlation manually but with new feature correlation is done for them

Reviewer 3: Kubeshark is adding hubs to monitor K8S API for example: https://docs.kubeshark.co/en/anatomy_of_kubeshark
Istio generates telemetry for meshes: https://istio.io/latest/docs/concepts/observability/
Managed kubernetes services like AKS / EKS does their own soup also. Not sure how it's implemented.

The point of view differs so they should be able to correlate / display differently.

This lack of granularity makes troubleshooting and optimization more challenging.

To solve this, we can:

Capture the network flows at the kernel level.

> Enrich them with translated endpoint information, showing not only the service but also the specific backend pod.

Reviewer 3: Could you please add a quick drawing explaining this ? I guess it will help :smile:

Author: the ui output with the svc traffic example not enough ?

Author: added new diagram to show how it works let me know what u think

Reviewer 3: thanks !

## How eBPF Can Help

You can execute custom programs using eBPF in the Linux kernel, making it an ideal tool for network observability.

Notable benefits of using eBPF include the following:

- Granular Observability: Directly see which pod served a request.

- Low Overhead: Operates in the kernel with minimal impact on performance.

- Dynamic Updates: Respond to changes in Kubernetes without requiring application restarts.

- Simplified Architecture: No need for sidecars or intrusive network plugins.

Here’s a high-level approach:

1- attach eBPF Programs:
Use eBPF programs to hook into kernel networking events, such as kprobe on functions like `nf_nat_manip_pkt`.
which will enable network observability eBPF agent to learn about all network translations events
done via [Linux Conntrack Tool](https://conntrack-tools.netfilter.org/manual.html)

2- enrich Flow Logs:
As network packets are processed, the eBPF hook will augment flow logs with metadata about the translated endpoint.

This will include:

- Source Pod IP

- Source Port

- Destination Pod IP

- Destination Port

> - Conntrack Zone ID

Reviewer 4: what is zone id?

Author: https://lwn.net/Articles/370152/#:~:text=A%20zone%20is%20simply%20a,to%20seperate%20conntrack%20defragmentation%20queues.

as two separate flows.
This lack of granularity makes troubleshooting and optimization more challenging.

To solve this, we can:

Capture the network flows at the kernel level.

Enrich them with translated endpoint information, showing not only the service but also the specific backend pod.
The Following diagram shows an example of kubernetes ClusterIP service and the translated endpoint information

<img src="{page.image('packets-xlation-enrichment/pkt-xlat-details.png')}" alt="Service Traffic's Packet Translation enrichment">

## How eBPF Can Help

You can execute custom programs using eBPF in the Linux kernel, making it an ideal tool for network observability.

Notable benefits of using eBPF include the following:

- Granular Observability: Directly see which pod served a request.

- Low Overhead: Operates in the kernel with minimal impact on performance.

- Dynamic Updates: Respond to changes in Kubernetes without requiring application restarts.

- Simplified Architecture: No need for sidecars or intrusive network plugins.

Here’s a high-level approach:

1- attach eBPF Programs:

> Use eBPF programs to hook into kernel networking events, such as kprobe on functions like `nf_nat_manip_pkt`.
> which will enable network observability eBPF agent to learn about all network translations events

Reviewer 5: nit: (there's a period in the middle of the sentence)

```suggestion
Use eBPF programs to hook into kernel networking events, such as kprobe on functions like `nf_nat_manip_pkt` which will enable network observability eBPF agent to learn about all network translation events
```
